1306_expansion_statements/expansion-statements.md

```diff
@@ -469,6 +469,409 @@ void f() {
+
+:::
+
+## 8.6+ [stmt.expand] Expansion statements
+
+Insert this section after [stmt.iter]{.sref} (and renumber accordingly).
+
+::: std
+::: addu
+**Expansion statements   [stmt.expand]**
+
+[1]{.pnum} Expansion statements specify repeated instantiations ([temp.spec]) of their substatement.
+
+```
+$expansion-statement$:
+    template for ( $init-statement$@~_opt_~@ $for-range-declaration$ : $expansion-initializer$ ) $statement$
+
+$expansion-initializer$:
+    $expression$
+    $expansion-init-list$
+
+$expansion-init-list$:
+    { $expression-list$ }
+```
+
+[Each `$decl-specifier$` in the `$decl-specifier-seq$` of a `$for-range-declaration$` is either a `$type-specifier$` or `constexpr` ([stmt.ranged]).]{.note}
+
+[#]{.pnum} The `$statement$` of an `$expansion-statement$` is a control-flow-limited statement ([stmt.label]).
+
+[#]{.pnum} For the purpose of name lookup and instantiation, the `$for-range-declaration$` and the `$statement$` of the `$expansion-statement$` are together considered a template definition.
+
+[#]{.pnum} An expansion statement whose `$expansion-initializer$` is an iterable expression ([stmt.ranged]) that does not have array type is an _iterable expansion statement_. An expansion statement that is not iterable and whose `$expansion-initializer$` is a non-type-dependent expression with a structured binding size ([dcl.struct.bind]) is a _destructurable expansion statement_. An expansion statement that is neither iterable nor destructurable shall have an `$expansion-initializer$` of the form `$expansion-init-list$` and is an _enumerated expansion statement_.
+
+[#]{.pnum} An expansion statement `$S$` with an `$expansion-initializer$` `$E$` is _expanded_ unless either
+
+- [#.#]{.pnum} `$E$` is type-dependent,
+- [#.#]{.pnum} `$S$` is an iterable expansion statement and `$E$` is value-dependent ([temp.dep.constexpr]) or
+- [#.#]{.pnum} `$S$` is an enumerated expansion statement and `$E$` contains a pack.
+
+[#]{.pnum} An expanded expansion statement `$S$` is equivalent to a `$compound-statement$` containing instantiations of the `$for-range-declaration$` (including its implied initialization), together with the `$statement$`; these instantiations correspond to expressions resulting from the analysis of the `$expansion-initializer$` as follows:
+
+- [#.#]{.pnum} If `$S$` is an iterable expansion statement, there is an instantiation for each element in the range computed by the `$expansion-initializer$`; `$S$` is equivalent to:
+  ```cpp
+  {
+    $init-statement$
+    static constexpr auto&& range = $expansion-initializer$ ;
+    // @i^th^@ repetition of the substatement
+    static constexpr auto iter@~_i_~@ = $get-expr$@~_i_~@ ;
```

Comment:
Not sure about `static` for either of these, is that still necessary after the references-in-structured-bindings paper? Maybe leave a note. 

Also, should the for-range-declaration allow `static`? I think there's an argument for yes, right?